[RFC] Warn on implicit float to int conversions.

Girgias opened the pull request:
Relatively straight forward implementation, might need to check OPcache.

RFC is located here: https://wiki.php.net/rfc/implicit-float-int-deprecate.

- [x] JIT Fixes
- [x] 32bit fixes

Girgias commented on Zend/tests/float_to_int/warnings_float_literals.phpt:
```diff
@@ -0,0 +1,147 @@
+}
+var_dump(foo(1.5));
+
+// TODO: Why are two warnings generated here?
+var_dump(chr(60.5));
```
Still don't understand why calls to internal functions trigger 2 diagnostics to be produced.

TysonAndre reviewed (commented):
I don't see any issues with this pr at a glance, only nits. I'm not familiar enough with opcache/jit to know if anything else needs to be changed.

I'd personally consider this an improvement over silently allowing the loss of precision, especially for array keys


Also, I think this will work with negative zero but it'd be useful to have at least one test case in case the implementation ever changes or the JIT ever unintentionally differs on some platforms. https://en.wikipedia.org/wiki/Signed_zero#Comparisons mentions

> According to the IEEE 754 standard, negative zero and positive zero should compare as equal with the usual (numerical) comparison operators, like the == operators of C and Java.

(i.e. since PHP's `===` uses C's `==` for floats I don't expect issues)

```
php > $negativeZero = -0.0;
php > var_dump($negativeZero);
float(-0)
php > var_dump($negativeZero === (float)(int)$negativeZero);
bool(true)
php > var_dump($negativeZero === 0.0);
bool(true)
```

TysonAndre commented on Zend/tests/float_to_int/explicit_casts_should_not_warn.phpt:
```diff
@@ -0,0 +1,34 @@
+int(0)
+int(3)
+int(3)
+int(9223372036854775807)
```
Should this test be skipped on 32-bit builds (`PHP_INT_SIZE === 4`)? Or maybe print PHP_INT_MAX as a string instead of calling var_dump if it would result in PHP_INT_MAX

See tests/lang/operators/bitwiseNot_basiclong_64bit.phpt

```
--SKIPIF--
<?php
if (PHP_INT_SIZE != 8) die("skip this test is for 64bit platform only");
?>
```

Girgias commented on Zend/zend_operators.h:
```diff
@@ -122,6 +122,7 @@ static zend_always_inline zend_long zend_dval_to_lval(double d)
 }
 #endif
 
+/* Used to convert a string float to integer during an (int) cast */
```
This is not used for floats only float strings, which explains the difference in behaviour with OOB floats and float strings in their integer representation.

Girgias commented on Zend/zend_operators.c:
```diff
@@ -807,6 +821,13 @@ ZEND_API zend_long ZEND_FASTCALL zval_get_long_func(zval *op) /* {{{ */
 		case IS_LONG:
 			return Z_LVAL_P(op);
 		case IS_DOUBLE:
+			if (EXPECTED(!is_lax)) {
```
Isn't that assumption only correct if ``ZEND_DVAL_TO_LVAL_CAST_OK`` is defined? (now it might always be the case but I haven't checked that)

nikic commented on Zend/zend_operators.h:
```diff
@@ -133,6 +134,18 @@ static zend_always_inline zend_long zend_dval_to_lval_cap(double d)
 }
 /* }}} */
 
+static zend_always_inline bool is_long_compatible(double d) {
```
zend_ prefix?

As written this code is also UB and will be flagged by ubsan. I think you may be better served by making this

```
static zend_always_inline bool zend_is_long_compatible(double dval, zend_long lval) {
    return (double) lval == dval;
}
```
and have the caller pass in the `zend_dval_to_lval` result. Also avoids a duplicate calculation.

nikic commented on Zend/zend_operators.h:
```diff
@@ -267,13 +280,16 @@ ZEND_API void ZEND_FASTCALL convert_to_boolean(zval *op);
 
 static zend_always_inline zend_long zval_get_long(zval *op) {
-	return EXPECTED(Z_TYPE_P(op) == IS_LONG) ? Z_LVAL_P(op) : zval_get_long_func(op);
+	return EXPECTED(Z_TYPE_P(op) == IS_LONG) ? Z_LVAL_P(op) : zval_get_long_func(op, 0);
```
This looks like a much larger change than what is described in the RFC. Doesn't this make all manual type conversions in extensions "strict"? We have always treated those the same way as explicit casts.

Girgias replied:
Oh I didn't know it was treated like a cast, will change this then again

nikic replied:
FWIW it might make sense to change that, but this probably isn't the place to start. Throwing a warning about float truncation while still happily and silently interpreting `[]` and `"foobar"` as `0` would be rather incongruent.

Girgias replied:
Yeah, but this would be an RFC to change the behaviour of userland casts I imagine?
I was considering anyway to make an (int) cast throw a notice/warning when casting a float larger than INT_MAX to align string float and normal float behaviour, but that's a rather separate change again.

nikic commented on Zend/Optimizer/sccp.c:
```diff
@@ -548,11 +558,16 @@ static inline int ct_eval_add_array_elem(zval *result, zval *value, zval *key) {
-		case IS_DOUBLE:
+		case IS_DOUBLE: {
+			zend_long lval = zend_dval_to_lval(Z_DVAL_P(key));
 			SEPARATE_ARRAY(result);
```
Move SEPARATE_ARRAY below the check.

nikic commented on Zend/zend_API.c:
```diff
@@ -501,22 +501,44 @@ ZEND_API bool ZEND_FASTCALL zend_parse_arg_long_weak(zval *arg, zend_long *dest,
+			/* Manually check arg_num is not (uint32_t)-1, as otherwise its called by
+			 * zend_verify_weak_scalar_type_hint_no_sideeffect() */
+			if (UNEXPECTED(!zend_is_long_compatible(Z_DVAL_P(arg), lval) && arg_num != (uint32_t)-1)) {
+				zend_error(E_DEPRECATED, "Implicit conversion to int from non-compatible float %f", Z_DVAL_P(arg));
```
This shouldn't be using `%f`, which uses limited precision. I think it's particularly important to do the print with maximum precision for these messages, because it may be exactly a difference at `10^(-16)` that makes the number not precisely representable.

I'd suggest doing the print with `%.*H` and `precision = -1`. You may want to exact functions for these two error messages to make sure they always use the same format.

It might also make sense to change the order a bit, `Implicit conversion from non-compatible float %f to int` would read more smoothly -- but I could also see an argument for your current phrasing, because it puts the variable value at the end of the message. No strong opinion either way.

nikic replied:
Now that there is a common function, you should also call it :D

Girgias replied:
How did I miss that... wonder if I should introduce a similar function for float-strings?

nikic commented on Zend/zend_operators.h:
```diff
@@ -267,13 +281,16 @@ ZEND_API void ZEND_FASTCALL convert_to_boolean(zval *op);
 ZEND_API void ZEND_FASTCALL convert_to_object(zval *op);
 
-ZEND_API zend_long    ZEND_FASTCALL zval_get_long_func(zval *op);
+ZEND_API zend_long    ZEND_FASTCALL zval_get_long_func(zval *op, bool is_lax);
```
Might make more sense to make this is_strict instead, otherwise most usages will be a double negation.

nikic commented on Zend/zend_operators.c:
```diff
@@ -792,7 +809,18 @@ ZEND_API void ZEND_FASTCALL convert_to_object(zval *op) /* {{{ */
+_Pragma("GCC diagnostic push")
+_Pragma("GCC diagnostic ignored \"-Wformat\"")
+_Pragma("clang diagnostic ignored \"-Wformat-invalid-specifier\"")
+_Pragma("GCC diagnostic ignored \"-Wformat-extra-args\"")
```
Ewww. I think you'd be better off adding zend_error_unchecked.

Girgias replied:
Yeah... this becoming pragma hell

nikic commented on Zend/zend_compile.c:
```diff
@@ -8306,7 +8306,7 @@ static bool zend_try_ct_eval_array(zval *result, zend_ast *ast) /* {{{ */
 				case IS_DOUBLE:
 					zend_hash_index_update(Z_ARRVAL_P(result),
-						zend_dval_to_lval(Z_DVAL_P(key)), value);
+						zend_dval_to_lval_safe(Z_DVAL_P(key)), value);
```
ct_eval shouldn't throw notices.

Girgias replied:
So how do I handle this? do I bailout?

nikic commented on Zend/tests/float_to_int/union_int_string_type_arg.phpt:
```diff
@@ -0,0 +1,28 @@
+--EXPECTF--
+int(1)
+
+Deprecated: Implicit conversion from non-compatible float 1.5 to int in %s on line %d
```
Interesting case. Would I be right in thinking that in PHP 9, the value is going to be (silently) treated as a string instead?

nikic commented on Zend/tests/float_to_int/warning_float_does_not_fit_zend_long_write_variation1.phpt:
```diff
@@ -0,0 +1,28 @@
+#0 {main}
+  thrown in %s on line %d
+
+Fatal error: Allowed memory size of %d bytes exhausted %S(tried to allocate %d bytes) in %s on line %d
```
Welp, this shouldn't happen. There's probably an exception check missing.

Girgias replied:
Indeed it's a pre-existing issue, openned #6890 to address it

nikic commented on Zend/tests/float_to_int/warning_float_does_not_fit_zend_long_write_variation2.phpt:
```diff
@@ -0,0 +1,28 @@
+--TEST--
+Implicit float to int conversions when float too large should warn, array variant
```
This doesn't look like an array variant.

nikic commented on Zend/tests/float_to_int/warnings_float_literals.phpt:
```diff
@@ -0,0 +1,139 @@
+
+Deprecated: Implicit conversion from non-compatible float 6.5 to int in %s on line %d
+
+Deprecated: Implicit conversion from non-compatible float 2.5 to int in %s on line %d
```
These deprecations shouldn't appear early. You need to handle this in zend_binary_op_produces_error.

nikic commented on Zend/tests/float_to_int/warnings_string_float_literals.phpt:
```diff
@@ -0,0 +1,97 @@
+
+Deprecated: Implicit conversion to int from non-compatible float-string 6.5 in %s on line %d
+
+Deprecated: Implicit conversion to int from non-compatible float-string 2.5 in %s on line %d
```
Same here, these warnings shouldn't be early.

nikic commented on Zend/zend_operators.c:
```diff
@@ -333,7 +342,15 @@ static zend_never_inline zend_long ZEND_FASTCALL zendi_try_get_long(zval *op, bo
-					 return zend_dval_to_lval_cap(dval);
+					lval = zend_dval_to_lval_cap(dval);
+					if (!zend_is_long_compatible(dval, lval)) {
+						zend_error(E_DEPRECATED, "Implicit conversion to int from non-compatible float-string %s",
```
```suggestion
						zend_error(E_DEPRECATED, "Implicit conversion from non-compatible float-string \"%s\" to int",
```
As the order for the other error message changed, I'd change it here as well.

nikic commented on Zend/tests/isset_str_offset.phpt:
```diff
@@ -60,7 +60,7 @@ var_dump(isset($str[$f]));
 
 ?>
---EXPECT--
+--EXPECTF--
```
Not sure about this one. Problem is that the current code just silently ignores all invalid types and returns false for them -- even the completely invalid ones like using an array as an offset. (Unlike the array case where this *does* result in a TypeError.)

nikic replied:
Though I do think that is *should* error, so I'm inclined to accept this...

nikic commented on Zend/tests/runtime_compile_time_binary_operands.phpt:
```diff
@@ -2,6 +2,7 @@
 Test binary operands exposing the same behavior at compile as at run time
 --INI--
 memory_limit=256M
+error_reporting=E_ALL & ~E_DEPRECATED
```
I believe this won't be necessary if you fix the compile-time evaluation.

nikic commented on Zend/zend_compile.c:
```diff
@@ -8332,10 +8332,16 @@ static bool zend_try_ct_eval_array(zval *result, zend_ast *ast) /* {{{ */
-						zend_dval_to_lval(Z_DVAL_P(key)), value);
+				case IS_DOUBLE: {
+					zend_long lval = zend_dval_to_lval(Z_DVAL_P(key));
+					/* TODO Should bailout on incompatible float?
```
Do what the `!zend_hash_next_index_insert` branch below does.

nikic commented on Zend/zend_execute.c:
```diff
@@ -775,16 +775,17 @@ static bool zend_verify_weak_scalar_type_hint_no_sideeffect(uint32_t type_mask,
 	bool bval;
 
-	if ((type_mask & MAY_BE_LONG) && zend_parse_arg_long_weak(arg, &lval, 0)) {
+	/* Pass (uint32_t)-1 as arg_num to indicate to ZPP not to emit any deprecation notice */
```
Not that it really matters, but as argument numbers are 1-based, you could have kept the 0 value ^^

Girgias commented:
@dstogov I tried fixing the JIT failures but I'm still stuck on the bitwise ops and modulo operators, could you have a look?

nikic commented:
Looks like there's one 32-bit failure left here.

Girgias commented:
I know, I need to go back to it, but I don't understand why this only happens on 32bit and not 64bit, as the value should also be out of integer range for 64bit (or maybe I made a mistake again)

nikic commented on Zend/Optimizer/zend_optimizer.c:
```diff
@@ -73,7 +73,7 @@ int zend_optimizer_eval_unary_op(zval *result, zend_uchar opcode, zval *op1) /*
 		if (opcode == ZEND_BW_NOT
 		 && Z_TYPE_P(op1) != IS_LONG
-		 && Z_TYPE_P(op1) != IS_DOUBLE
+		 /* && Z_TYPE_P(op1) != IS_DOUBLE */ // TODO Optimize when float is int compatible
```
I've changed this to use zend_unary_op_produces_error() in https://github.com/php/php-src/commit/ba9c0b38cb141767d99e41789bc96c336608262b, so the change can be dropped from here.

nikic commented:
> I know, I need to go back to it, but I don't understand why this only happens on 32bit and not 64bit, as the value should also be out of integer range for 64bit (or maybe I made a mistake again)

It looks correct to me. On 64-bit `-2147483649` will be an integer, on 32-bit it will be a float.

Girgias commented on Zend/zend_compile.c:
```diff
@@ -8217,10 +8229,10 @@ static inline bool zend_try_ct_eval_binary_op(zval *result, uint32_t opcode, zva
 {
 	if (opcode == ZEND_BW_NOT) {
-		return Z_TYPE_P(op) <= IS_TRUE || Z_TYPE_P(op) == IS_ARRAY;
+		return (Z_TYPE_P(op) <= IS_TRUE || Z_TYPE_P(op) == IS_ARRAY || Z_TYPE_P(op) == IS_DOUBLE);
```
Yes, I think I didn't do it initially as I was trying to figure out how to get consistent behaviour with Opcache/JIT and when it's disabled.

nikic commented:
> > > I know, I need to go back to it, but I don't understand why this only happens on 32bit and not 64bit, as the value should also be out of integer range for 64bit (or maybe I made a mistake again)
> > 
> > 
> > It looks correct to me. On 64-bit `-2147483649` will be an integer, on 32-bit it will be a float.
> 
> I changed this value in the test to `-9_223_372_036_854_775_900` which _should_ be less than the minimum value on 64 bit (-9,223,372,036,854,775,808)

It is -- but floats only have 54 bits precision so the float literal itself will already have lost precision and have the same value as `(float)PHP_INT_MIN`. You need to go significantly below that to hit a different float :)

Girgias commented:
> > > > I know, I need to go back to it, but I don't understand why this only happens on 32bit and not 64bit, as the value should also be out of integer range for 64bit (or maybe I made a mistake again)
> > > 
> > > 
> > > It looks correct to me. On 64-bit `-2147483649` will be an integer, on 32-bit it will be a float.
> > 
> > 
> > I changed this value in the test to `-9_223_372_036_854_775_900` which _should_ be less than the minimum value on 64 bit (-9,223,372,036,854,775,808)
> 
> It is -- but floats only have 54 bits precision so the float literal itself will already have lost precision and have the same value as `(float)PHP_INT_MIN`. You need to go significantly below that to hit a different float :)

Ah yes... changed the value to something lower then and removing the XFAIL, just crossing fingers nothing else fails after the refactoring.

nikic commented on Zend/zend_compile.c:
```diff
@@ -8156,6 +8156,27 @@ static bool zend_try_ct_eval_magic_const(zval *zv, zend_ast *ast) /* {{{ */
+ZEND_API bool zend_is_op_long_compatible(zval *op)
+{
+	/* This is handled before */
+	ZEND_ASSERT(Z_TYPE_P(op) != IS_ARRAY);
```
It looks like there's a pre-existing issue here:
```
<?php
$x = 1;
$x <<= [];
```
This does not throw a TypeError with opcache enabled. I think you should just return false for IS_ARRAY here and that add test case.

Girgias replied:
Shouldn't this be fixed in 7.4/8.0 too if it's an Opcache bug?

nikic commented on Zend/zend_compile.c:
```diff
@@ -8207,6 +8228,12 @@ ZEND_API bool zend_binary_op_produces_error(uint32_t opcode, zval *op1, zval *op
+	/* Operation which cast float/float-strings to integers might produce incompatible float to int errors */
+	if (opcode == ZEND_SL || opcode == ZEND_SR || opcode == ZEND_BW_OR
+			|| opcode == ZEND_BW_AND || opcode == ZEND_BW_XOR || opcode == ZEND_MOD) {
+		return (!zend_is_op_long_compatible(op1) || !zend_is_op_long_compatible(op2));
```
```suggestion
		return !zend_is_op_long_compatible(op1) || !zend_is_op_long_compatible(op2);
```
Redundant parens

nikic commented on Zend/zend_compile.c:
```diff
@@ -8226,7 +8253,7 @@ static inline bool zend_try_ct_eval_binary_op(zval *result, uint32_t opcode, zva
 {
 	if (opcode == ZEND_BW_NOT) {
-		return Z_TYPE_P(op) <= IS_TRUE || Z_TYPE_P(op) == IS_ARRAY;
+		return (Z_TYPE_P(op) <= IS_TRUE || Z_TYPE_P(op) == IS_ARRAY || !zend_is_op_long_compatible(op));
```
With the modification above, the IS_ARRAY check here can be dropped.

nikic commented on Zend/zend_operators.h:
```diff
@@ -133,6 +134,22 @@ static zend_always_inline zend_long zend_dval_to_lval_cap(double d)
 /* }}} */
 
+static zend_always_inline bool zend_is_long_compatible(double d, zend_long l) {
+	return ((double)l == d);
```
```suggestion
	return (double)l == d;
```
Redundant parens

nikic commented on Zend/zend_vm_def.h:
```diff
@@ -9028,6 +9028,7 @@ ZEND_VM_COLD_CONST_HANDLER(190, ZEND_COUNT, CONST|TMPVAR|CV, UNUSED)
 				zval retval;
 
 				zend_call_method_with_0_params(zobj, NULL, NULL, "count", &retval);
+				// TODO Should this use zval_get_long_ex(&retval, /* is_strict */ false);
```
That's what it uses now :)

nikic commented on ext/mysqli/tests/mysqli_stmt_bind_result_format.phpt:
```diff
@@ -227,7 +227,7 @@ memory_limit=83886080
                 do {
                     $trend = (mt_rand(0, 3) > 1) ? (mt_rand(-10000, 10000) / 100) : 'NULL';
-                } while (isset($values[$trend]));
+                } while (isset($values[(int)$trend]));
```
I think in this case `(string)$trend` might have been intended, as `$trend` is used as a float value here.

nikic commented on ext/standard/array.c:
```diff
@@ -6052,7 +6052,8 @@ PHP_FUNCTION(array_key_exists)
 			break;
 		case IS_DOUBLE:
-			RETVAL_BOOL(zend_hash_index_exists(ht, zend_dval_to_lval(Z_DVAL_P(key))));
+			/* TODO Should this warn or not? */
```
As we also warn here for resource, and also throw the deprecation in isset(), I believe we so.

Girgias replied:
Wouldn't it make sense to use a union type here? If I understand the behaviour correctly we would need `int|string|null` as `null` is evaluated as `''` instead of `0`, in which case we get the handling of the other types for free.

nikic replied:
As the conversion semantics are supposed to match those of normal array access, I think it's clearer not to use union types. If nothing else, we'd get a difference in error level for the resource case.

Girgias replied:
ACK, will handle manually then.

nikic commented on ext/standard/formatted_print.c:
```diff
@@ -634,15 +634,14 @@ php_formatted_print(char *format, size_t format_len, zval *args, int argc, int n
-										  zval_get_long(tmp),
-										  width, padding, alignment,
-										  always_sign);
+						zval_get_long_ex(tmp, /* is_strict */ true),
```
I'm not sure about these changes. I could see an argument being made that `%d` should be considered an explicit int cast.

Girgias replied:
Well `%s` does emit a warning when you pass in an array so I thought throwing the deprecation notice would be somewhat consistent (https://3v4l.org/hfbbX)

nikic replied:
This warning comes from the explicit cast: https://3v4l.org/Frq47

Girgias replied:
TIL

nikic commented on ext/standard/tests/array/array_intersect_assoc_variation5.phpt:
```diff
@@ -67,8 +67,10 @@ foreach($arrays as $arr1) {
+--EXPECTF--
 *** Testing array_intersect_assoc() : assoc array with diff keys to $arr1 argument ***
+
+Deprecated: Implicit conversion from non-compatible float 444.432 to int in %s on line %d
```
This deprecation warning is thrown in an array literal for the test data. In such cases, I think it would be better to remove the element from the array / adjust the key, rather than adding the deprecation warning to the output. Otherwise there will be more work when this becomes an error.

The same also applies to some tests after this.

nikic commented:
ext/standard/tests/strings/vprintf_variation15.phpt fails on 32-bit.

derickr commented:
Are you going to bring this up for a vote at some point soon?

Girgias commented:
> Are you going to merge this at some point soon?

Hopefully during next week, now that most of the issues have been solved.

nikic commented on Zend/zend_vm_def.h:
```diff
@@ -5230,7 +5230,7 @@ ZEND_VM_C_LABEL(send_array):
 			uint32_t skip = opline->extended_value;
 			uint32_t count = zend_hash_num_elements(ht);
-			zend_long len = zval_get_long(op2);
+			zend_long len = zval_get_long_ex(op2, /* is_strict */ true);
```
This should be using zpp ... fixed in https://github.com/php/php-src/commit/27be6670b11cbe909a27d51c8e31590fc9283079.

nikic commented on ext/standard/tests/strings/chunk_split_variation10.phpt:
```diff
@@ -45,36 +45,66 @@ for($count = 0; $count < count($values); $count++) {
 *** Testing chunk_split() : different single quoted strings as 'ending' ***
 -- Iteration 0 --
+
+Deprecated: Implicit conversion from non-compatible float 9.2 to int in %s on line %d
```
This tests looks useless, let's delete it.

nikic commented on ext/standard/tests/strings/fprintf_variation_002.phpt:
```diff
@@ -3,7 +3,7 @@ Test fprintf() function (variation - 2)
 <?php
 
-$int_variation = array( "%d", "%-d", "%+d", "%7.2d", "%-7.2d", "%07.2d", "%-07.2d", "%'#7.2d" );
+$int_variation = array( '"%d"', '"%-d"', '"%+d"', '"%7.2d"', '"%-7.2d"', '"%07.2d"', '"%-07.2d"', '"%\'#7.2d"' );
```
Why this change?

Girgias replied:
These changes were made because of trailing whitespaces, I think they are useful to keep but probably as a separate PR/commit.

nikic commented on ext/standard/tests/strings/fprintf_variation_003.phpt:
```diff
@@ -7,7 +7,7 @@ if (PHP_INT_SIZE != 4) die("skip this test is for 32bit platform only");
 <?php
 
-$int_numbers = array( 0, 1, -1, 2.7, -2.7, 23333333, -23333333, "1234" );
+$int_numbers = array( 0, 1, -1, 23333333, -23333333, "1234" );
```
Probably these changes to printf tests should be dropped now.

nikic commented:
Nearly there...

Girgias commented on ext/standard/tests/math/decbin_variation1_64bit.phpt:
```diff
@@ -10,61 +10,31 @@ if (PHP_INT_SIZE != 8) die("skip this test is for 64bit platform only");
-       18446744073709551615,  // largest decimal
-       18446744073709551616,
+       4294967295,  // largest decimal
+       4294967296,
```
Probably a copy/paste error

Ayesh commented:
Thank you @Girgias for this, and congratulations on the unanimous RFC vote yes.

Really trying not to bike-shed, but do you think using `incompatible` instead of `non-compatible` would be any better? Google [ngram](https://books.google.com/ngrams/graph?content=incompatible%2Cnon-compatible&year_start=1900&year_end=2000&corpus=0&smoothing=50&case_insensitive=true#) shows `incompatible` is used more than `non-compatible`, and perhaps we could improve the error message a little bit more?

We already use `incompatible` in error messages, but I don't think we do for other error messages apart from this one.

If you think it's a good idea, I would gladly send a PR through. I apologize if the choice of `non-compatible` is deliberate; English is my secondary language and `incompatible` felt a bit more natural, hence this comment. 

Thank you.

Girgias commented:
Feel free to send a PR, I didn't think a lot about the deprecation message to be honest (and it's been a while). :-)

Ayesh commented:
Thanks a lot @Girgias - I just made #7095 with the changes.
